test/SILOptimizer/escape_analysis.sil

```diff
@@ -1310,3 +1311,91 @@ bb(%0 : $X, %1: $@callee_owned (X, X) -> (@out X, @error ErrorType)):
   %7 = tuple()
 	return %7 : $()
 }
+
+// CHECK-LABEL: CG of check_dealloc_ref
+// CHECK-NEXT:   Arg %0 Esc: A, Succ:
+// CHECK-NEXT:  End
+sil @check_dealloc_ref : $@convention(thin) (@owned X) -> () {
+bb0(%0 : $X):
+  dealloc_ref %0 : $X
+  %1 = tuple ()
+  return %1 : $()
+}
+
+// CHECK-LABEL: CG of check_set_deallocating
+// CHECK-NEXT:   Arg %0 Esc: A, Succ:
+// CHECK-NEXT:  End
+sil @check_set_deallocating : $@convention(thin) (@owned X) -> () {
+bb0(%0 : $X):
+  set_deallocating %0 : $X
+  %1 = tuple ()
+  return %1 : $()
+}
+
+// Escape analysis should analyze the implcit destructor call which may
+// happen due to strong_release and figure out that the object will not
+// globally escape.
+// CHECK-LABEL: CG of check_implicit_destructor_invocation_via_strong_release
+// CHECK-NEXT:   Val %0 Esc: %1, Succ:
+// CHECK-NEXT:  End
+sil @check_implicit_destructor_invocation_via_strong_release : $@convention(thin) () -> () {
+bb0:
+  %0 = alloc_ref [stack] $X
+  strong_release %0 : $X
+  dealloc_ref [stack] %0 : $X
+  %1 = tuple ()
+  return %1 : $()
+}
+
+// Escape analysis should analyze the implcit destructor call which may
+// happen due to strong_release and figure out that the object will not
+// globally escape.
+// CHECK-LABEL: CG of check_implicit_destructor_invocation_via_release_value
+// CHECK-NEXT:   Val %0 Esc: %1, Succ:
+// CHECK-NEXT:  End
+sil @check_implicit_destructor_invocation_via_release_value : $@convention(thin) () -> () {
+bb0:
+  %0 = alloc_ref [stack] $X
+  release_value %0 : $X
+  dealloc_ref [stack] %0 : $X
+  %1 = tuple ()
+  return %1 : $()
+}
+
+// Check that escape analysis can look through bb arguments and figure
+// out that all objects are local. By analyzing a destructor implicitly
+// invoked by strong_release it would also determine that these objects
+// do not escape from the function.
+// CHECK-LABEL: CG of check_is_local_object_through_bb_args
+// CHECK-LABEL:  Val %2 Esc: %6,%7, Succ: 
+// CHECK-LABEL:  Val %4 Esc: %6,%7, Succ: 
+// CHECK-LABEL:  Val %6 Esc: %6,%7, Succ: %2, %4
+// CHECK-LABEL: End
+sil @check_is_local_object_through_bb_args: $@convention(thin) (Builtin.Int1) -> () {
+bb0(%0 : $Builtin.Int1):
+  cond_br %0, bb1, bb2
+
+bb1:
+  %2 = alloc_ref $X
+  br bb3(%2 : $X)
+
+bb2:
+  %5 = alloc_ref $X
+  br bb3(%5 : $X)
+
+bb3(%8: $X):
+  strong_release %8 : $X
+  %9 = tuple ()
+  return %9 : $()
+}
```

Comment:
@eeckstein I added a couple of tests to demonstrate that destructors which allow objects to escape are handled properly.